pkg/endpoint/bpf.go

```diff
@@ -302,22 +302,20 @@ func (e *Endpoint) addNewRedirectsFromDesiredPolicy(ingress bool, desiredRedirec
 				direction = trafficdirection.Egress
 			}
 
-			keysFromFilter := l4.ToMapState(e, direction, &policyIdentitiesLabelLookup{e})
-
+			idLookup := &policyIdentitiesLabelLookup{e}
+			keysFromFilter := l4.ToMapState(e, direction, idLookup)
 			for keyFromFilter, entry := range keysFromFilter {
 				if oldEntry, ok := e.desiredPolicy.PolicyMapState[keyFromFilter]; ok {
 					// Keep the original old entry for revert if there are duplicate keys,
 					// which are possible due to named ports resolving to the same number.
 					if _, isDup := updatedDesiredMapState[keyFromFilter]; !isDup {
 						updatedDesiredMapState[keyFromFilter] = oldEntry
 					}
-				} else {
-					insertedDesiredMapState[keyFromFilter] = struct{}{}
 				}
 				if entry.IsRedirectEntry() {
 					entry.ProxyPort = redirectPort
 				}
-				e.desiredPolicy.PolicyMapState[keyFromFilter] = entry
+				e.desiredPolicy.PolicyMapState.DenyPreferredInsertWithChanges(keyFromFilter, entry, insertedDesiredMapState, nil, idLookup)
```

Comment:
Is it expected that `DenyPreferredInsertWithChanges` will insert into `insertedDesiredMapState`?

Reply:
Previously we were inserting into `insertedDesiredMapState` only when the key/entry did not exist previously, now `DenyPreferredInsertWithChanges` inserts into `insertedDesiredMapState` also when the there was a previous entry with the same key. This does not matter, as where these are used in the revert code below, we first delete all inserted keys, then add back the old values, if they existed, from `updatedDesiredMapState`. End result is that same before and after, previously the revert code directly replaced the new value with the old one, now the new value is deleted first, and the old one added back after.

Reply:
Hmm. if there were old values also for (e.g., L3L4) entries created by `DenyPreferredInsertWithChanges` then the old values of those can not be reverted back in case of a failure with the current code in the PR. I may have to change `adds` and `deletes` arguments to `DenyPreferredInsertWithChanges` to be `MapState` instead of just `Keys` to make `DenyPreferredInsertWithChanges` revertible. Let me try that!